packages/shared/src/permissions/permissions.utils.ts

```diff
@@ -31,3 +31,77 @@ export function hasPermission(
 export function roleSupportsEnvLimit(role: MemberRole): boolean {
   return ["engineer", "experimenter"].includes(role);
 }
+
+export type ReadAccessFilter = {
+  globalReadAccess: boolean;
+  projects: { id: string; readAccess: boolean }[];
+};
+
+export function getReadAccessFilter(userPermissions: UserPermissions) {
+  const readAccess: ReadAccessFilter = {
+    globalReadAccess: userPermissions.global.permissions.readData || false,
+    projects: [],
+  };
+
+  Object.keys(userPermissions.projects).forEach((project) => {
+    readAccess.projects.push({
+      id: project,
+      readAccess:
+        userPermissions.projects[project].permissions.readData || false,
+    });
+  });
+
+  return readAccess;
+}
+
+export function hasReadAccess(
+  readAccessFilter: ReadAccessFilter,
+  resourceProjects: string[] // feature.project or experiment.project or metric.projects
+): boolean {
```

Comment:
This function is really hard to follow.  This much nesting and branching is a little suspicious. Does the logic need to be this complicated?

Reply:
Just took a pass at updating this to remove a lot of the nesting. Still not super easy to follow, but better than the previous iteration.